### PR TITLE
Add support for gpio-poweroff

### PR DESCRIPTION
**Pull request by theophile**
See discussion at https://forum.radxa.com/t/gpio-poweroff-in-linux/10644

### CODE REVIEW

File: drivers/power/reset/gpio-poweroff.c [hunk 1]
```diff
@@ -76,14 +80,21 @@ static int gpio_poweroff_probe(struct platform_device *pdev)
 	if (IS_ERR(reset_gpio))
 		return PTR_ERR(reset_gpio);
 
+	export = of_property_read_bool(pdev->dev.of_node, "export");
```
Comment (RadxaYuntian) [inline drivers/power/reset/gpio-poweroff.c:84]: This block appears to come from [this commit](https://github.com/raspberrypi/linux/commit/fbaddd65c8e5d588ffee8dc8b5b7b9faa8434387). However, you missed the [variable declaration](https://github.com/raspberrypi/linux/commit/fbaddd65c8e5d588ffee8dc8b5b7b9faa8434387#diff-ac6fa903566521bf39bc66f8de1995a3ad1b9faec23be1b459cad52997464e61R34). I double this commit can build now.

**Review by RadxaYuntian — CHANGES_REQUESTED** [review]
Please rewrite the history to match Raspberry Pi's [commit log](https://github.com/raspberrypi/linux/commits/rpi-5.15.y/drivers/power/reset/gpio-poweroff.c). Currently this seems broken.

File: arch/arm64/boot/dts/amlogic/overlay/meson-g12a-gpio-poweroff.dts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+                                /* Sets poweroff pin to GPIOAO_9 as an example. Modify as needed. */
+                                gpios = <&gpio_ao GPIOAO_9 GPIO_ACTIVE_LOW>;
+		                timeout-ms = <20000>;
+				force;
+		        };
+		};
+	};
+};
+
+
```
Comment (RadxaYuntian) [inline arch/arm64/boot/dts/amlogic/overlay/meson-g12a-gpio-poweroff.dts]: Please also remove the extra empty line.

**Comment by RadxaYuntian** [comment]
I applied the following patch and the overlay. However, when I shorted pin 37 & 39, nothing happens. I could see `/sys/firmware/devicetree/base/gpio-poweroff` and `/sys/bus/platform/drivers/poweroff-gpio` so the driver is included and the overlay is loaded.

[Archive.zip](https://github.com/radxa/kernel/files/9746230/Archive.zip)

**Comment by theophile** [comment]
> I applied the following patch and the overlay. However, when I shorted pin 37 & 39, nothing happens. I could see `/sys/firmware/devicetree/base/gpio-poweroff` and `/sys/bus/platform/drivers/poweroff-gpio` so the driver is included and the overlay is loaded.
> 
> [Archive.zip](https://github.com/radxa/kernel/files/9746230/Archive.zip)

That is correct. gpio-poweroff is an output pin that is normally high, but when the system has completed its shutdown process, the [gpio-poweroff pin is pulled low momentarily](https://www.kernel.org/doc/Documentation/devicetree/bindings/power/reset/gpio-poweroff.txt). This is to signal an external power source that it is now safe to remove power. To test if it's working, you can connect an led or multimeter leads with (+) on pin 37 and (-) on pin 39. The LED should be on (or +3.3V showing) while the system is running. Then issue `shutdown now -h` and when the shutdown is complete, the LED should turn off (or multimeter drop to 0V) momentarily before going back on again.

File: drivers/power/reset/gpio-poweroff.c (hunk 1 above)
Comment (theophile) [reply drivers/power/reset/gpio-poweroff.c:83]: You're right. Sorry about that.

Comment (theophile) [reply drivers/power/reset/gpio-poweroff.c:84]: Added a commit that fixes this.

File: arch/arm64/boot/dts/amlogic/overlay/meson-g12a-gpio-poweroff.dts (hunk 2 above)
Comment (theophile) [reply arch/arm64/boot/dts/amlogic/overlay/meson-g12a-gpio-poweroff.dts]: Done.

**Comment by RadxaYuntian** [comment]
@RadxaStephen Please rebuild the kernel for Zero.

**Pull request merged**